sdk/core/switcher/entry.S

```diff
@@ -504,207 +831,348 @@ exception_entry_asm:
-	// This may result in something out-of-bounds if the compartment has a
-	// malicious value for their error handler (hopefully caught at link or
-	// load time), but if it does then we will double-fault and force unwind.
+	/*
+	 * Set the jump target to the error handler entry point.  This may result in
+	 * something out-of-bounds if the compartment has a malicious value for
+	 * their error handler (hopefully caught at link or load time), but if it
+	 * does then we will fault when attempting the cjalr below and force unwind
+	 * (either because the cjalr itself will raise a fault, because ra is
+	 * untagged, or because the resulting PCC is out of bounds and instruction
+	 * fetch fails; either case results in a forced unwind, albeit by slightly
+	 * different paths, with .Lhandle_error_switcher_pcc relevant for the former
+	 * and .Lhandle_error_test_double_fault for the latter.
+	 */
 	cgetbase           s1, cra
 	csetaddr           cra, cra, s1
 	cincoffset         cra, cra, s0
 
-	// If we're in an error handler with a stack, set up the stack, otherwise
-	// we just need to set up argument registers.
+	/*
+	 * If we're in an error handler with a stack, set up the stack, otherwise
+	 * we just need to set up argument registers.
+	 */
+.Lhandle_error_test_stackful:
+	// FROM: above
 	beqz               a0, .Lset_up_stack_handler
+
+.Lset_up_stack_handler_stackless:
+	// FROM: above
 	clw                a0, TrustedStack_offset_mcause(csp)
 	csrr               a1, mtval
 	li                 a2, 0
 	cmove              csp, ct0
+	// Atlas: sp: taget compartment invocation stack pointer
 	j                  .Linvoke_error_handler
 
 .Lset_up_stack_handler:
-	// Set up the on-stack context for the callee
-	clc                cs1, 0(csp)
+	/*
+	 * FROM: .Lhandle_error_test_stackful
+	 * LIVE IN: ra, sp, gp
+	 *
+	 * Atlas:
+	 *  ra: handler entrypoint (with bounds of compartment's .text)
+	 *  sp: pointer to TrustedStack
+	 *  gp: target compartment cgp
+	 */
+	/*
+	 * Set up the on-stack context, a compartment.h:/struct ErrorState value,
+	 * which has the same layout at a TrustedStack spill frame.
+	 *
+	 * These begin with a PCC.  To ensure that handlers do not have access to
+	 * values (especially, capabilities) reachable through the trapping PCC,
+	 * we clear the tag.  Handlers of course retain access to values reachable
+	 * through their own PCC and CGP.
+	 */
+	clc                cs1, TrustedStack_offset_mepcc(csp)
 	ccleartag          cs1, cs1
-	csc                cs1, 0(ct0)
-	// Source for context copy.
-	cincoffset         ca2, csp, TrustedStack_offset_cra
-	// Destination for context copy
-	cincoffset         ca3, ct0, TrustedStack_offset_cra
-	copyContext        ca3, ca2, cs1, a4
+	csc                cs1, TrustedStack_offset_mepcc(ct0)
+	// Now copy the 15 GPRs from the trusted stack (sp)
+	cincoffset         ca2, csp, TrustedStack_offset_cra // source
+	cincoffset         ca3, ct0, TrustedStack_offset_cra // destination
```

Comment:
Why did you change the comment style here?  We don't put comments on the end of lines anywhere else in the switcher, do we?